### PR TITLE
OJ-2735 implements dequeue lambda and audit events table for test harness

### PR DESCRIPTION
**Pull request by Caitlin-cooling**
### What changed

New test harness stack introduced including:
- Audit events table to store audit events
- Dequeue lambda and related infrastructure to receive messages from audit event queue and put them into the audit events table
- Tests and linting config

[Common-infra PR](https://github.com/govuk-one-login/ipv-cri-common-infrastructure/pull/99) to export the AuditEventQueueArn

### Why did it change

This is a long term solution for testing txma audit events

### Screenshots
Screenshot 2024-08-30 at 10 26 49 AM
Screenshot 2024-08-30 at 10 26 56 AM
Screenshot 2024-08-30 at 10 27 36 AM

### Issue tracking

- [OJ-2735](https://govukverify.atlassian.net/browse/OJ-2735)

### Environment variables or secrets

- [x] Documented in the [README](./blob/main/README.md)
- [x] Added to deployment repository
- [x] Added to local startup repository

### Other considerations

- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

[OJ-2735]: https://govukverify.atlassian.net/browse/OJ-2735?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[OJ-2738]: https://govukverify.atlassian.net/browse/OJ-2738?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test-harness/infrastructure/samconfig.toml
Comment (Caitlin-cooling) [inline test-harness/infrastructure/samconfig.toml:1]: Open to discussion about this - I find this the easiest way create config for deployment but happy to put in a deploy.sh if people prefer

File: test-harness/lambdas/src/dequeue-handler.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        for await (const { body, messageId } of event.Records) {
+            const { sessionId, eventName, timestamp } = this.getEventData(body);
+            const nowInSeconds = Math.floor(Date.now() / 1000);
+            const ttl = nowInSeconds + 360;
+
+            const putItemCommand: PutItemCommand = new PutItemCommand({
+                TableName: tableName,
+                Item: {
+                    partitionKey: { S: `SESSION#${sessionId}` },
```
Comment (Caitlin-cooling) [inline test-harness/lambdas/src/dequeue-handler.ts:24]: If anyone can think of better names please shout!

File: test-harness/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Test Harness
+
+This is a test harness that listens to events from the TxMA SQS queue and puts them a test events table which can then be accessed using an API to allow automated testing of TxMA events.
```
Comment (jkunle) [inline test-harness/README.md:3]: `puts them into`

File: test-harness/infrastructure/template.yaml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        AWS_STACK_NAME: !Sub ${AWS::StackName}
+        POWERTOOLS_LOG_LEVEL: DEBUG
+        POWERTOOLS_METRICS_NAMESPACE: !Ref CriIdentifier
+    AutoPublishAlias: live
+
+Resources:
+  DequeueFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+      FunctionName: !Sub "Audit-Event-Dequeue-${AWS::StackName}"
```
Comment (jkunle) [inline test-harness/infrastructure/template.yaml:76]: Recently we try not to name resources, as it is good practice to let SAM do this for you. Unless there is a very good reason to. It helps with resource name lengths. I know, the common lambda template doesn't really follow this (it's old).

File: test-harness/infrastructure/samconfig.toml
Comment (jkunle) [reply test-harness/infrastructure/samconfig.toml:1]: I don't have any significant objection, apart from being consistent. If we use the above then we shouldn't forget to add appropriate tags as well see https://github.com/govuk-one-login/ipv-cri-common-lambdas/blob/main/deploy.sh#L27-L30

**Comment by jkunle** [comment]
Am sure you were not aware, we have the https://github.com/govuk-one-login/ipv-cri-templates this would have created a lot of the bootstrap code you've had to do here. You just need to point to a folder it will scaffold the template and directory structure, reading for you, it does readme, linting stuff and can scaffold lambda test for you as well.

**Review by gds-jbentham — DISMISSED** [review]

File: test-harness/README.md (hunk 2 above)
Comment (Caitlin-cooling) [reply test-harness/README.md]: Nice spot done

File: test-harness/infrastructure/template.yaml (hunk 3 above)
Comment (Caitlin-cooling) [reply test-harness/infrastructure/template.yaml]: Ah good to know - I've removed the name from this and the table

File: test-harness/infrastructure/samconfig.toml
Comment (Caitlin-cooling) [reply test-harness/infrastructure/samconfig.toml:1]: Cool done

File: test-resources/test-harness/lambdas/src/dequeue-handler.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+export const dbClient = new DynamoDBClient({ region: process.env.REGION });
+
+export class DequeueLambdaHandler implements LambdaInterface {
+    async handler(event: SQSEvent): Promise<{ batchItemFailures: SQSBatchItemFailure[] }> {
+        logger.info("Starting to process records");
+
+        const tableName = process.env.EVENTS_TABLE_NAME;
+        const batchItemFailures: SQSBatchItemFailure[] = [];
+
+        for await (const { body, messageId } of event.Records) {
```
Comment (jkunle) [inline test-resources/test-harness/lambdas/src/dequeue-handler.ts:16]: await not sure it's need here  think it this one that's important `await dbClient.send(putItemCommand);`

**Review by jkunle — DISMISSED** [review]
This looks good

**Review by jkunle — APPROVED** [review]

**Pull request merged**